File: src/main/kotlin/io/javalin/plugin/rendering/template/JavalinStringTemplate4.java
package io.javalin.plugin.rendering.template;

import io.javalin.http.Context;
import io.javalin.rendering.FileRenderer;
import io.javalin.rendering.JavalinRenderer;
import org.jetbrains.annotations.NotNull;
import org.stringtemplate.v4.ST;
import org.stringtemplate.v4.STGroup;
import org.stringtemplate.v4.STRawGroupDir;

import java.util.Map;
import java.util.function.Consumer;

public class JavalinStringTemplate4 implements FileRenderer {

    private static final char defaultDelimiter = '$';
    private static final String defaultTemplateDir = "templates/st";
    private static final String defaultModelAccessor = "m";
    private static String modelAccessor = "m";
    private static STRawGroupDir group;

    public static void init(Consumer<JavalinStringTemplate4Configuration> consumer) {
        JavalinRenderer.register(new JavalinStringTemplate4(), ".st", ".html.st");
        JavalinStringTemplate4Configuration config = new JavalinStringTemplate4Configuration();
        consumer.accept(config);
        STGroup.verbose = config.verbose;
        group = new STRawGroupDir(config.templateDir, "UTF-8", config.delimiter, config.delimiter);
        group.load();
        modelAccessor = config.modelAccessor;
    }

    @Override
    public String render(@NotNull String filePath, @NotNull Map<String, Object> model, Context context) throws Exception {
        ST template = group.getInstanceOf(cleanFilePath(filePath));
        template.add(modelAccessor, model);
        return template.render();
    }

    private String cleanFilePath(String filePath) {
        return filePath.endsWith(".st") ? filePath.substring(0, filePath.length() - 3) : filePath;
    }

    public static class JavalinStringTemplate4Configuration {

        private String templateDir = defaultTemplateDir;
        private String modelAccessor = defaultModelAccessor;
        private char delimiter = defaultDelimiter;
        private boolean verbose = false;

        public void templateDir(String templateDir) {
            this.templateDir = templateDir;
        }

        public void modelAccessor(String modelAccessor) {
            this.modelAccessor = modelAccessor;
        }

        public void delimiter(char delimiter) {
            this.delimiter = delimiter;
        }

        public void verbose(boolean verbose) {
            this.verbose = verbose;
        }
    }

}

File: src/main/java/io/javalin/rendering/template/JavalinStringTemplate4.kt
package io.javalin.rendering.template

import io.javalin.http.Context
import io.javalin.rendering.FileRenderer
import io.javalin.rendering.JavalinRenderer
import io.javalin.rendering.JavalinRenderer.register
import io.javalin.rendering.util.RenderingDependency.STRING_TEMPLATE_4
import io.javalin.rendering.util.Util.throwIfNotAvailable
import org.stringtemplate.v4.STGroup
import org.stringtemplate.v4.STRawGroupDir
import java.util.function.Consumer

class JavalinStringTemplate4 : FileRenderer {

    override fun render(filePath: String, model: Map<String, Any?>, ctx: Context): String {
        if (reloadOnRender) {
            group!!.unload()
            group!!.load()
        }
        val template = group!!.getInstanceOf(cleanFilePath(filePath))
        model.forEach(template::add)
        return template.render()
    }


    private fun cleanFilePath(filePath: String): String {
        return if (filePath.endsWith(".st")) filePath.substring(0, filePath.length - 3) else filePath
    }

    class JavalinStringTemplate4Configuration {
        var templateDir = defaultTemplateDir
        var devMode = false;
        var delimiter = defaultDelimiter

        fun templateDir(templateDir: String) {
            this.templateDir = templateDir
        }

        fun devMode(devMode: Boolean) {
            this.devMode = devMode;
        }

        fun delimiter(delimiter: Char) {
            this.delimiter = delimiter
        }


    }

    companion object {
        private const val defaultDelimiter = '$'
        private const val defaultTemplateDir = "templates/st"
        private var reloadOnRender = false
        private var verbose = false
        private var group: STRawGroupDir? = null

        val extensions = arrayOf(".st")

        @JvmStatic
        @JvmOverloads
        fun init(consumer: Consumer<JavalinStringTemplate4Configuration>? = null) {
            throwIfNotAvailable(STRING_TEMPLATE_4)
            register(JavalinStringTemplate4(), *extensions)
            val config = JavalinStringTemplate4Configuration()
            consumer?.accept(config)
            verbose = config.devMode
            reloadOnRender = config.devMode
            STGroup.verbose = verbose
            group = STRawGroupDir(config.templateDir, "UTF-8", config.delimiter, config.delimiter)
            group!!.load()
        }
    }

    class Loader : JavalinRenderer.FileRendererLoader {
        override fun load() = if (!JavalinRenderer.hasRenderer(*extensions) && STRING_TEMPLATE_4.exists()) init() else Unit
    }

}
